host/frontend/host-orchestrator/instancemanager.go

```diff
@@ -33,3 +56,110 @@ func validateRequest(r *CreateCVDRequest) error {
 	}
 	return nil
 }
+
+type FetchCVDHandler struct {
+	dir                string
+	fetchCVDDownloader FetchCVDDownloader
+	mutex              sync.Mutex
+}
+
+func NewFetchCVDHandler(dir string, fetchCVDDownloader FetchCVDDownloader) *FetchCVDHandler {
+	return &FetchCVDHandler{dir, fetchCVDDownloader, sync.Mutex{}}
+}
+
+func (h *FetchCVDHandler) Download(buildID, accessToken string) error {
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
+	exist, err := h.exist(buildID)
+	if err != nil {
+		return err
+	}
+	if !exist {
+		fileName := h.buildFileName(buildID)
+		f, err := os.Create(fileName)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+		if err := h.fetchCVDDownloader.Download(f, buildID, accessToken); err != nil {
+			if err := os.Remove(fileName); err != nil {
+				return err
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+func (h *FetchCVDHandler) exist(buildID string) (bool, error) {
+	if _, err := os.Stat(h.buildFileName(buildID)); err == nil {
+		return true, nil
+	} else if errors.Is(err, os.ErrNotExist) {
+		return false, nil
+	} else {
+		return false, err
+	}
+}
+
+func (c *FetchCVDHandler) buildFileName(buildID string) string {
+	return fmt.Sprintf("%s/fetch_cvd_%s", c.dir, buildID)
+}
+
+type FetchCVDDownloader interface {
+	Download(dst io.Writer, buildID string, accessToken string) error
+}
+
+// Downloads fetch_cvd binary from Android Build.
+type ABFetchCVDDownloader struct {
```

Comment:
this is more about having a relevant name different than the interface name, and less about downloading fetch_cvd from another place/system. please, let me know if you have another name more suitable, happy to change it.

Reply:
I guess it could stay that way. The other option is to name the interface CVDFetcherDownloader and the impl FetchCVDDownloader since the ab/ specific command is fetch_cvd and the generic name could be CVDFetcher. I don't have a strong preference either way.